python/tank/descriptor/bundle_cache_usage/manager.py

```diff
@@ -0,0 +1,270 @@
+        MAX_DEPTH_WALK = 2
+
+        bundle_path_list = []
+        bundle_cache_root = self.bundle_cache_root
+
+        # Process the local app store first
+        bundle_cache_app_store = BundleCacheManager._find_app_store_path(bundle_cache_root)
+
+        if bundle_cache_app_store:
+            log.debug("Found local app store path: %s" % (bundle_cache_app_store))
+
+            if bundle_cache_app_store and \
+                    os.path.exists(bundle_cache_app_store) and \
+                    os.path.isdir(bundle_cache_app_store):
+
+                log.debug("Found local app store path: %s" % (bundle_cache_app_store))
+                bundle_path_list += BundleCacheManager._find_descriptors(bundle_cache_app_store, MAX_DEPTH_WALK)
+        else:
+            log.debug("Could not find the local app store path from: %s" % (bundle_cache_root))
+
+        return bundle_path_list
+
+    def _find_bundles(self):
+        """
+        Scan the bundle cache at the specified location for bundles and add them
+        as unused entries to the database.
+
+        Reference: Walk up to a certain level
+        https://stackoverflow.com/questions/42720627/python-os-walk-to-certain-level
+
+        :param bundle_cache_root: A str of a path
+        """
+
+        app_store_bundle_list = self._find_app_store_bundles()
+        # TODO: Process other bundle_cache sub folders
+        bundle_list2 = []
+        bundle_list3 = []
+
+        # Combine all lists into a single one
+        bundle_path_list = list(set(app_store_bundle_list + bundle_list2 + bundle_list3))
+
+        return bundle_path_list
+
+    @classmethod
+    def _get_filelist(cls, bundle_path):
+
+        if not os.path.exists(bundle_path):
+            raise BundleCacheUsageException(bundle_path, "The specified path does not exists.")
+
+        file_list = []
+        for (dirpath, dirnames, filenames) in os.walk(bundle_path):
+            file_list.append(dirpath)
+            for filename in filenames:
+                fullpath = os.path.join(dirpath, filename)
+                file_list.append(fullpath)
+
+        return file_list
+
+    def _paranoid_delete(self, filelist):
+        """
+
+        Delete files and folder under the specified filelist in a paranoid mode where
+        everything is carrefully checked before deleteion. That means no 'rmtree'-like
+        operation, no walking and deleting items directly.
+
+        On anything unexpected the process stops with a custom exception.
+
+        We cannot delete file right away. The list being in reverse order, we might be
+        trying to delete a file that exists in a symlinked folder. Therefore, we'll first
+        scan the entire list for a link, symlink or such in the list. If found, we'll
+        abort the deletion process before actually deleting anything.
+
+        :param filelist: A list file and folders to be deleted
+        """
+
+        # First, check whether there is a symlink in the list
```

Comment:
for I/O, we should use the methods in `util.filesystem` unless there is a good reason not to!

Reply:
Why are we reimplementing shutil? It does symlink validation and everything. Symlinks are files and shutil will not visit symlinks even if it points to a folder. I just tried it just to confirm.

Reply:
The rationale behind this apparent wheel re-invented comes from a requirement mentionned by @manneohrstrom. He mentionned that we should delete files individually so we can double check we're not deleting anything unwanted, or trying to delete locked file (Mostly Windows I guess): no rmtree delete. Also, prior to starting the deletion process, we would check for symlink presence, abort deletion process before even deleting anything and finaly report an error on deleting a given 'bundle' 

I believe those we're not possible out of the box with existing file management utility methods hence having my own version called: paranoid_delete ( in reference to the famous unix-cd-ripping-utility reading mode). Again, I did mention that in the note I wrote 15 days ago. Without any comment on the note, I assumed everyone was fine with its content.

Reply:
I think Manne and I disagree here on the role shtuil.rmtree can play here. You can add a callback to shutil that handles errors like file locking and what not. We could simply log the error and move on to the next file or raising an exception and catch it on our code.

Reply:
@manneohrstrom , what do you think?